### ISSUE TITLE
Proxies i programowanie na typach przy tworzeniu ergonomicznych i tree shakable bibliotek

### ISSUE THREAD

**Issue by mieszkosabo** [meetjspl/poznan#96]
Cześć, 
tytuł jest trochę dziwny, ale ostatnio myślałem nad tym czy da się robić takie biblioteki jak np. Zod, ale tak, by user mógł sam sobie wybrać podzbiór API, który go interesuje i dzięki temu zmniejszyć bundle size swojeje aplikacji.

Tak to wygląda w Zodzie:

```ts
import { z } from "zod"; // 12.8 kB

const LoginSchema = z.object({
  email: z.string().email(),
  password: z.string().min(8),
});
```

Nawet jeśli potrzebujemy tylko walidatorów do `string`, `email` i `min`, to i tak musimy zaimportować całe `z` ważące 12.8 kB.

No i okazuje się, że wymyśliłem jak to zrobić i o tym chciałbym opowiedzieć. Zrobiłem taką proof-of-concept libkę, która używa tej techniki [correttore](https://github.com/mieszkosabo/correttore):

```ts
import { email, minLength, object, initCorrettore, string } from "correttore"; // 0.54 kB

export const c = initCorrettore({
  string,
  email,
  object,
  minLength,
});

const LoginSchema = c.object({
  email: c.string().email(),
  password: c.string().minLength(8),
});
```

Dajcie znać, czy to brzmi ciekawie. Wiem, że jest spotkanie 10.10, jeśli jest jeszcze miejsce, to chętnie wpadnę, a jak nie, to może następnym razem.

**Comment by SebastianPozoga** [comment]
Brzmi bardzo ciekawie. Ale na najbliższe spotkanie mamy komplet, więc prezentacja pójdzie na następne. Co ciekawe zebrały nam się już 3 prezentacje, więc nie będzie trzeba długo czekać :-)

**Comment by mieszkosabo** [comment]
Fajnie, to czekam na info :)

**Comment by SebastianPozoga** [comment]
Cześć
Czy pasuje Ci termin 14.11.2023 na poprowadzenie prezentacji?

**Comment by mieszkosabo** [comment]
Tak

**Comment by pawk3k** [comment]
https://github.com/fabian-hiller/valibot

**Comment by mieszkosabo** [comment]
@pawk3k Yep, talk będzie dotyczył tego jak połączyć "najlepsze z obu światów", czyli stworzyć libkę, która jest tree shakable jak valibot, ale jednocześnie ma nice, chainable API jak zod.

I oczywiście biblioteka do walidacji schemy, to tylko przykład dla ustalenia uwagi. Myślę, że technika ma więcej use caseów.

**Comment by naugtur** [comment]
Chciałbym to zobaczyć. Będzie nagranie? :sweat_smile:

**Comment by SebastianPozoga** [comment]
postaram się tematu przypilnować lepiej niż ostatnio :-)
